File: app/src/main/java/com/example/learningverbs/utils/CustomListEventListener.java
package com.example.learningverbs.utils;

import androidx.annotation.NonNull;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;

import java.util.ArrayList;

public abstract class CustomListEventListener<T> implements ValueEventListener {
    final ArrayList<T> verbList = new ArrayList<>();
    private Class<T> customClass;

    public CustomListEventListener(Class<T> GenericTypeIndicator) {
        showLoaging();
        this.customClass = GenericTypeIndicator;
    }

    @Override
    public void onDataChange(@NonNull DataSnapshot snapshot) {
        hideLoading();
        for (DataSnapshot postSnapshot : snapshot.getChildren()) {
            verbList.add(postSnapshot.getValue(customClass));
        }
        if (verbList.isEmpty()) {
            onFailed(new Throwable("Error al obtener los datos"));
        } else {
            onSuccess(verbList);
        }

    }

    @Override
    public void onCancelled(@NonNull DatabaseError error) {
        hideLoading();
        onFailed(new Throwable(error.getMessage()));
    }


    public abstract void onSuccess(ArrayList<T> response);
    public abstract void onFailed(Throwable throwable);
    public abstract void showLoaging();
    public abstract void hideLoading();

}

File: app/src/main/java/com/karla/learningverbs/kotlin/utils/firebase/CustomListEventListener.kt
package com.karla.learningverbs.kotlin.utils.firebase

import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
abstract class CustomListEventListener<T>(GenericTypeIndicator: Class<T>) :
    ValueEventListener {
    val verbList = ArrayList<T?>()
    private val customClass: Class<T>

    init {
        showLoaging()
        customClass = GenericTypeIndicator
    }

    override fun onDataChange(snapshot: DataSnapshot) {
        hideLoading()
        verbList.clear()
        for (postSnapshot in snapshot.children) {
            verbList.add(postSnapshot.getValue(customClass))
        }
        if (verbList.isEmpty()) {
            onFailed(Throwable("Error al obtener los datos"))
        } else {
            onSuccess(verbList)
        }
    }

    override fun onCancelled(error: DatabaseError) {
        hideLoading()
        onFailed(Throwable(error.message))
    }

    abstract fun onSuccess(response: ArrayList<T?>?)
    abstract fun onFailed(throwable: Throwable?)
    abstract fun showLoaging()
    abstract fun hideLoading()
}